src/mongoose_rabbit_worker.erl

```diff
@@ -133,12 +139,34 @@ declare_exchange(Channel, Exchange) ->
 delete_exchange(Channel, Exchange) ->
     amqp_channel:call(Channel, #'exchange.delete'{exchange = Exchange}).
 
--spec publish_status(Status :: atom(), Channel :: pid(), Exchange :: binary(),
-                     JID :: binary()) -> term().
-publish_status(Status, Channel, Exchange, JID) ->
-    Message = presence_msg(JID, Status),
+-spec publish_status(Channel :: pid(), map()) -> term().
+publish_status(Channel, #{user_jid := JID,
+                          exchange := Exchange,
+                          status := Status}) ->
+    Message = make_presence_msg(JID, Status),
     publish(Channel, Exchange, JID, Message).
 
+-spec publish_chat_msg_sent(Channel :: pid(), EventData :: map()) -> ok.
+publish_chat_msg_sent(Channel, EventData = #{from_jid := From,
+                                             topic := Topic}) ->
+    RoutingKey = user_topic_routing_key(From, Topic),
+    publish_chat_msg_event(Channel, RoutingKey, EventData).
+
+-spec publish_chat_msg_received(Channel :: pid(), EventData :: map()) -> ok.
+publish_chat_msg_received(Channel, EventData = #{to_jid := To,
+                                                 topic := Topic}) ->
+    RoutingKey = user_topic_routing_key(To, Topic),
+    publish_chat_msg_event(Channel, RoutingKey, EventData).
+
+-spec publish_chat_msg_event(Channel :: pid(), RoutingKey :: binary(), map()) ->
+    term().
+publish_chat_msg_event(Channel, RoutingKey, #{from_jid := From,
```

Comment:
Im not sure if this helper is necessary. It just wraps 2 function calls. I think such decoupling is not helpful, because without this function, code it really readable and code duplication is minimal.
```
    RoutingKey = user_topic_routing_key(From/To, Topic),
    Message = make_chat_msg(From, To, UserMsg),
    publish(Channel, Exchange, RoutingKey, Message).
```

Also, after removal we would have consistent flow for all events:

```
+---------------------------------------+        +----------------------+
|                                       |        |                      |
| handle_cast(user_presence_changed, _) +-------->  publish_status(...) +---------------------------+
|                                       |        |                      |                           |
+---------------------------------------+        +----------------------+                           |
                                                                                                    |
                                                                                                    |
+---------------------------------------+        +-----------------------------+         +----------v-------+
|                                       |        |                             |         |                  |
| handle_cast(user_chat_msg_sent, _)    +-------->  publish_chat_msg_sent(...) +--------->   publish(...)   |
|                                       |        |                             |         |                  |
+---------------------------------------+        +-----------------------------+         +----------^-------+
                                                                                                    |
                                                                                                    |
+---------------------------------------+        +-----------------------------+                    |
|                                       |        |                             |                    |
| handle_cast(user_chat_msg_recv, _)    +-------->  publish_chat_msg_rec (...) +--------------------+
|                                       |        |                             |
+---------------------------------------+        +-----------------------------+

```